### PR TITLE
Visual Studio for Mac Updates

### PR DESCRIPTION
**Pull request by Saratsin**
### Description:
Huge updates for Visual Studio for Mac:

- "Format Xaml" context menu button in editor is back
- "Format Xaml" context menu button on document tab
- "Format Xaml on save" option is working now
- "Search to drives root" option is working now
- "External configuration file" option is working now 

Fixes issues: #196 #199 
Probably fixes issues: #201 #231 

### Checklist:
* [x] I have commented my code, particularly in hard-to-understand areas
* [x] My changes generate no new warnings

### CODE REVIEW

File: XamlStyler.Mac/Gui/XamlStylerOptionsPanel.cs [hunk 1]
```diff
@@ -1,19 +1,26 @@
...
-		private OptionsViewModel _optionsViewModel = new OptionsViewModel();
+        public XamlStylerOptionsPanel()
+        {
+            ViewModel = new XamlStylerOptionsViewModel();
+        }
 
-		public override MonoDevelop.Components.Control CreatePanelWidget()
+        private XamlStylerOptionsViewModel ViewModel { get; }
+
+		public override Control CreatePanelWidget()
```
Comment (NicoVermeir) [inline XamlStyler.Mac/Gui/XamlStylerOptionsPanel.cs]: This code is not aligned with the rest of the file

File: XamlStyler.Mac/Gui/XamlStylerOptionsWidget.cs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+using Xavalon.XamlStyler.Mac.Utils;
+using Gtk;
+using Xavalon.XamlStyler.Mac.ViewModels;
+
+namespace Xavalon.XamlStyler.Mac.Gui
+{
+	[ToolboxItem(true)]
+	public partial class XamlStylerOptionsWidget : Bin
+	{
+		private readonly XamlStylerOptionsViewModel _viewModel;
```
Comment (NicoVermeir) [inline XamlStyler.Mac/Gui/XamlStylerOptionsWidget.cs]: align the code in this file

File: XamlStyler.Mac/Gui/XamlStylerOptionsWidget.cs [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+		private readonly Color _groupHeaderColor;
+		private readonly Color _altRowColor;
+
+		public XamlStylerOptionsWidget(XamlStylerOptionsViewModel viewModel)
+		{
+			_viewModel = viewModel;
+
+            if (MonoDevelop.Ide.IdeApp.Preferences.UserInterfaceTheme == MonoDevelop.Ide.Theme.Dark)
+            {
+                _groupHeaderColor = ColorUtils.Parse("#222222");
```
Comment (NicoVermeir) [inline XamlStyler.Mac/Gui/XamlStylerOptionsWidget.cs]: can these hard-coded values be stored in variables somewhere?

File: XamlStyler.Mac/ViewModels/XamlStylerOptionViewModel.cs [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+using System.ComponentModel;
+using Xavalon.XamlStyler.Core.Options;
+
+namespace Xavalon.XamlStyler.Mac.ViewModels
+{
+    public class XamlStylerOptionViewModel
+	{
+		public XamlStylerOptionViewModel(PropertyDescriptor property)
+		{
+            var browsableAttribute = (BrowsableAttribute)property.Attributes[typeof(BrowsableAttribute)];
```
Comment (NicoVermeir) [inline XamlStyler.Mac/ViewModels/XamlStylerOptionViewModel.cs]: align code

File: XamlStyler.Mac/Gui/XamlStylerOptionsPanel.cs (hunk 1 above)
Comment (Saratsin) [reply XamlStyler.Mac/Gui/XamlStylerOptionsPanel.cs]: should be aligned now

File: XamlStyler.Mac/Gui/XamlStylerOptionsWidget.cs (hunk 2 above)
Comment (Saratsin) [reply XamlStyler.Mac/Gui/XamlStylerOptionsWidget.cs]: should be aligned now

File: XamlStyler.Mac/Gui/XamlStylerOptionsWidget.cs (hunk 3 above)
Comment (Saratsin) [reply XamlStyler.Mac/Gui/XamlStylerOptionsWidget.cs]: moved it to separate Colors class

File: XamlStyler.Mac/ViewModels/XamlStylerOptionViewModel.cs (hunk 4 above)
Comment (Saratsin) [reply XamlStyler.Mac/ViewModels/XamlStylerOptionViewModel.cs]: should be aligned now

**Comment by Saratsin** [comment]
Fixed all comments

File: XamlStyler.Mac/Properties/AddinInfo.cs [hunk 5]
```diff
@@ -1,7 +1,7 @@
 ﻿using Mono.Addins;
 using Mono.Addins.Description;
 
-[assembly: Addin("XamlStyler", Namespace = "Xavalon", Version = "1.1.5")]
+[assembly: Addin("XamlStyler", Namespace = "Xavalon", Version = "3.2001.0")]
```
Comment (grochocki) [inline XamlStyler.Mac/Properties/AddinInfo.cs]: Can we continue with the old versioning scheme here for now? Technically, 3.2001.0 already shipped. The goal of #231 is to automate versioning and support simultaneous releases.

File: XamlStyler.Mac/Properties/AddinInfo.cs [hunk 6]
```diff
@@ -1,7 +1,7 @@
...
 
 [assembly: AddinName("XAML Styler")]
 [assembly: AddinCategory("IDE extensions")]
```
Comment (grochocki) [inline XamlStyler.Mac/Properties/AddinInfo.cs]: We publish all binaries under "Xavalon" and drop our personal affiliations. Can you update the `AddinAuthor` below to reflect that?

File: XamlStyler.Mac/Properties/AssemblyInfo.cs [hunk 7]
```diff
@@ -17,7 +16,7 @@
 // The form "{Major}.{Minor}.*" will automatically update the build and revision,
 // and "{Major}.{Minor}.{Build}.*" will update just the revision.
 
-[assembly: AssemblyVersion("1.1.*")]
+[assembly: AssemblyVersion("3.2001.*")]
```
Comment (grochocki) [inline XamlStyler.Mac/Properties/AssemblyInfo.cs]: Revert versioning schema change here too

File: XamlStyler.Mac/CommandHandlers/BatchFormatXamlCommandHandler.cs [hunk 8]
```diff
@@ -0,0 +1,75 @@
+﻿using MonoDevelop.Components.Commands;
```
Comment (grochocki) [inline XamlStyler.Mac/CommandHandlers/BatchFormatXamlCommandHandler.cs]: Can you add our standard header to the top of all code files?

"// © Xavalon. All rights reserved."

File: XamlStyler.Mac/Properties/Manifest.addin.xml [hunk 9]
```diff
@@ -1,36 +1,44 @@
...
         <Condition id="ItemType" value="Project|Solution">
-            <CommandItem id="Xavalon.XamlStyler.Mac.FormatXamlBatch" />
+            <CommandItem id="Xavalon.XamlStyler.Mac.BatchFormatXaml" />
         </Condition>
     </Extension>
+    
     <Extension path="/MonoDevelop/Ide/Commands/Edit">
-        <Command id="Xavalon.XamlStyler.Mac.FormatXaml" _label="Format Xaml" _description="Reformats the XAML document to be much more readable" defaultHandler="Xavalon.XamlStyler.Mac.FormatXamlHandler" macShortcut="Meta|Shift|I" />
-        <Command id="Xavalon.XamlStyler.Mac.FormatXamlBatch" _label="Batch Format Xaml" _description="Reformats all XAML documents to be much more readable" defaultHandler="Xavalon.XamlStyler.Mac.FormatXamlBatchHandler" />
+        <Command
```
Comment (grochocki) [inline XamlStyler.Mac/Properties/Manifest.addin.xml]: Does this expose two options for the user to select between?

Comment (Saratsin) [reply XamlStyler.Mac/Properties/Manifest.addin.xml]: These commands are quite different in behavior.

FormatXaml is command for formatting single Xaml document. It appears in 3 places:
- Text Editor => Right Click => Context Menu => Format Xaml
- Text Editor => File Tab => Right Click => Context Menu => Format Xaml
- Toolbar => Edit Menu => Format Xaml

FormatXamlBatch is command for formatting all Xaml files inside specific project/solution. It appears in 2 places:
- Solution Pad => Project Item => Right Click => Batch Format Xaml
- Solution Pad => Solution Item => Right Click => Batch Format Xaml

**Comment by Saratsin** [comment]
### **Commands UI**

**Format Xaml** (format single xaml file):
1. Text Editor => Right Click => Context Menu => Format Xaml
image

2. Text Editor => File Tab => Right Click => Context Menu => Format Xaml
image

3. Toolbar => Edit Menu => Format Xaml
image


**Batch Format Xaml** (formatting all xaml files inside specific project/solution):
1. Solution Pad => Project Item => Right Click => Batch Format Xaml
image

2. Solution Pad => Solution Item => Right Click => Batch Format Xaml
image

### **Settings UI**
image

File: XamlStyler.Mac/Properties/AddinInfo.cs (hunk 5 above)
Comment (Saratsin) [reply XamlStyler.Mac/Properties/AddinInfo.cs]: Okay, I've changed it to 2.0.0, it has some breaking changes, including renaming of global options file storage so it's major update

File: XamlStyler.Mac/Properties/AddinInfo.cs (hunk 6 above)
Comment (Saratsin) [reply XamlStyler.Mac/Properties/AddinInfo.cs]: Left only Xavalon as AddinAuthor

File: XamlStyler.Mac/Properties/AssemblyInfo.cs (hunk 7 above)
Comment (Saratsin) [reply XamlStyler.Mac/Properties/AssemblyInfo.cs]: Reverted

File: XamlStyler.Mac/Properties/Manifest.addin.xml (hunk 9 above)
Comment (Saratsin) [reply XamlStyler.Mac/Properties/Manifest.addin.xml]: Added explanation with screenshots in message below

File: XamlStyler.Mac/CommandHandlers/BatchFormatXamlCommandHandler.cs (hunk 8 above)
Comment (Saratsin) [reply XamlStyler.Mac/CommandHandlers/BatchFormatXamlCommandHandler.cs]: Added copyright everywhere

File: XamlStyler.Mac/XamlStyler.Mac.csproj [hunk 10]
```diff
@@ -79,7 +79,7 @@
   <ItemGroup>
     <PackageReference Include="Mono.Addins" Version="1.3.8" />
     <PackageReference Include="MonoDevelop.Addins" Version="0.4.7" />
-    <PackageReference Include="Newtonsoft.Json" Version="10.0.3" />
+    <PackageReference Include="Newtonsoft.Json" Version="12.0.3" />
```
Comment (grochocki) [inline XamlStyler.Mac/XamlStyler.Mac.csproj]: Is 12.0.3 required? If possible, we should stay consistent between projects, and we cannot move off of 9.0.1 until we drop support for VS2017 (more details in #243).

**Comment by grochocki** [comment]
> Batch Format Xaml (formatting all xaml files inside specific project/solution)

Really cool! This looks like #56. Whenever possible, we should avoid any divergence between Windows and Mac versions when it comes to features. At the same time, this is an approved feature... so some options:
1) Limit this PR to bug fixes and split off batch processing in a separate PR that implements #56 for both Windows and Mac
2) Add batch processing to Windows version too as part of this PR
3) Take this PR as is, referencing #56, but now #56 _must_ be included in vNext.

Thoughts? Until we unify releases with #231, I open to any of these, but I would generally prefer (1) or (2).

File: XamlStyler.Mac/XamlStyler.Mac.csproj (hunk 10 above)
Comment (Saratsin) [reply XamlStyler.Mac/XamlStyler.Mac.csproj]: Unfortunately it's required. Package not compiling if we have lower version (see screenshot):
image

I guess Visual Studio or MonoDevelop has a dependency on Newtonsoft.Json. Before update of Visual Studio for Mac I had the same error, but it was asking for version 10.0.0.0, after I updated Visual Studio for Mac and MonoDevelop to v8.4, it started to write about version 12.0.0.0

Comment (grochocki) [reply XamlStyler.Mac/XamlStyler.Mac.csproj]: Ah, gotcha. No worries then. The way Visual Studio and Newtonsoft versioning are coupled was an interesting rabbit hole I crawled down to understand some odd errors on Windows, but I think the take away was that we should just use the minimum version that works. It doesn't conflict with anything else in the project, so we should be good.

**Comment by Saratsin** [comment]
This feature (Batch Format Xaml) was present in VS for Mac since its first release (it was present even before I started support it. I guess it was implemented by Ben Lacey). I think 1 is bad option, because some users on Mac might already know about this feature and use it.

I can take the implementation of #56 for Windows (also maybe will try to unify the logic between two extensions), but I'll be able to start it only after 12 of January (until that moment I'll not have access to the Windows PC).

My proposal is next: take option 3 and I'll start doing feature #56

**Comment by grochocki** [comment]
Sounds good to me, thanks!

**Pull request merged**